engine/immediate_task_conrrent_processor.go

```diff
@@ -186,14 +186,24 @@ func (w *immediateTaskConcurrentProcessor) processWaitUntilTask(
 		defer httpResp.Body.Close()
 	}
 	if w.checkResponseAndError(err, httpResp) {
-		status, details, err := w.composeHttpError(err, httpResp, prep.Info, task)
+		status, details, _ := w.composeHttpError(err, httpResp, prep.Info, task)
 
 		nextIntervalSecs, shouldRetry := w.checkRetry(task, prep.Info)
 		if shouldRetry {
 			return w.retryTask(ctx, task, prep, nextIntervalSecs, status, details)
 		}
-		// TODO otherwise we should fail the state and process execution if the backoff is exhausted, unless using a recovery policy
-		return err
+
+		errRecover := w.applyStateFailureRecoveryPolicy(ctx,
+			task,
+			prep,
+			status,
+			details,
+			task.ImmediateTaskInfo.WorkerTaskBackoffInfo.CompletedAttempts,
+			xdbapi.WAIT_UNTIL_API)
+		if errRecover != nil {
+			return errRecover
+		}
+		return nil
```

Comment:
looks like it can be just 
```golang
return w.applyStateFailureRecoveryPolicy(ctx,
			task,
			prep,
			status,
			details,
			task.ImmediateTaskInfo.WorkerTaskBackoffInfo.CompletedAttempts,
			xdbapi.WAIT_UNTIL_API)

```